src/signed.ml

```diff
@@ -62,8 +65,93 @@ struct
   let (asr) = shift_right
 end
 
+module type Small = sig
+  val bits : int
+
+  val min : unit -> int
+  val max : unit -> int
+  val of_string : string -> int
+  val to_string : int -> string
+  val to_hexstring : int -> string
+end
+
+module MakeSmall(S : Small) =
+struct
+  open S
+
+  let fix i = (i lsl (int_size - bits)) asr (int_size - bits)
+
+  module Basics =
+  struct
+    type t = int
+    external to_int : t -> int = "%identity"
+    let min_int = S.min ()
+    let max_int = S.max ()
+    let of_string = of_string
+    let to_string = to_string
+    let to_hexstring = to_hexstring
+
+    let add : t -> t -> t = fun x y -> fix (x + y)
+    let sub : t -> t -> t = fun x y -> fix (x - y)
+    let mul : t -> t -> t = fun x y -> fix (x * y)
+    let div : t -> t -> t = ( / )
+    let rem : t -> t -> t = ( mod )
+    let logand : t -> t -> t = ( land )
+    let logor : t -> t -> t = ( lor )
+    let logxor : t -> t -> t = ( lxor )
+    let shift_left : t -> int -> t = fun x y -> fix (x lsl y)
+    let shift_right : t -> int -> t = ( asr )
+    let shift_right_logical : t -> int -> t = fun x y ->
+      ((x lsl (int_size - bits)) lsr y) asr (int_size - bits)
```

Comment:
This is done to handle the case of preservation when the RHS is 0 -- it's a bit ugly so if there's a better way to do it I'm all ears.